kong/plugins/grpc-gateway/deco.lua

```diff
@@ -51,165 +67,207 @@ local valid_method = {
-  if not protofile then
-    return nil
-  end
+--- Return input and output names of the method specified by the url path
+-- @param method
+-- @param path
+-- @param cfg
+-- @return
+-- TODO: memoize - still valid?
+local function rpc_transcode( method, path, cfg )
+  local info = get_call_info( cfg )
 
-  local info = get_proto_info(protofile)
-  info = info[method]
-  if not info then
-    return nil, ("Unknown method %q"):format(method)
+  local endpoints = info.endpoints[ method ] or nil
+
+  if not endpoints then
+    return nil, ("unknown method %q"):format(method)
   end
-  for _, endpoint in ipairs(info) do
-    local m, err = re_match(path, endpoint.regex, "jo")
+
+  for _, endpoint in ipairs( endpoints ) do
+    local m, err = re_match( path, endpoint.regex, "jo" )
+
     if err then
-      return nil, ("Cannot match path %q"):format(err)
+      return nil, ("cannot match path %q"):format(err)
     end
+
     if m then
       local vars = {}
-      for i, name in ipairs(endpoint.varnames) do
-        vars[name] = m[i]
+      for i, name in ipairs( endpoint.varnames ) do
+        -- check handling of dotted variables (e.g. entity.uuid)
+        vars[ name ] = m[ i ]
       end
       return endpoint, vars
     end
   end
-  return nil, ("Unknown path %q"):format(path)
-end
 
+  return nil, ("unknown path %q"):format(path)
+end
 
-function deco.new(method, path, protofile)
-  if not protofile then
+--- Constructor
+-- @param method
+-- @param path
+-- @param cfg
+function deco.new( method, path, cfg )
+  if not cfg.proto then
     return nil, "transcoding requests require a .proto file defining the service"
   end
 
-  local endpoint, vars = rpc_transcode(method, path, protofile)
+  local endpoint, vars = rpc_transcode( method, path, cfg )
 
   if not endpoint then
     return nil, "failed to transcode .proto file " .. vars
   end
 
+  local info = get_call_info( cfg )
+
+  protojson:configure( {
+    use_proto_names = cfg.use_proto_names,
+    enum_as_name = cfg.enum_as_name,
+    emit_defaults = cfg.emit_defaults,
+    json_names = info.json_names,
+  })
+
   return setmetatable({
     template_payload = vars,
     endpoint = endpoint,
     rewrite_path = endpoint.rewrite_path,
   }, deco)
 end
 
-local function get_field_type(typ, field)
-  local _, _, field_typ = pb.field(typ, field)
-  return field_typ
-end
-
-local function encode_fix(v, typ)
-  if typ == "bool" then
-    -- special case for URI parameters
-    return v and v ~= "0" and v ~= "false"
-  end
-
-  return v
-end
-
---[[
-  // Set value `v` at `path` in table `t`
-  // Path contains value address in dot-syntax. For example:
-  // `path="a.b.c"` would lead to `t[a][b][c] = v`.
-]]
-local function add_to_table( t, path, v, typ )
-  local tab = t -- set up pointer to table root
-  local msg_typ = typ;
-  for m in re_gmatch( path , "([^.]+)(\\.)?") do
-    local key, dot = m[1], m[2]
-    msg_typ = get_field_type(msg_typ, key)
```

Comment:
Are we removing the check? Is this a valid change?
(Could it be possible for a gRPC endpoint to receive args irrelevant to the gRPC call?

Reply:
The way it is done now, [only fields from schema are included](https://github.com/Kong/kong/blob/20c2d67780ea36931edadb43aa195a6b1cc90a87/kong/tools/protojson.lua#L410) so such a field will not make it into proto.

Officially this behavior [should be controlled](https://github.com/protocolbuffers/protobuf-go/blob/v1.30.0/encoding/protojson/decode.go#L41), but not now, not here :)